### ISSUE TITLE
Add short(ish) timeouts for HTTP for interactive operations

### ISSUE THREAD

**Issue by larsks** [restic/restic#1713]
## Output of `restic version`

restic 0.8.3
compiled with go1.10 on linux/amd64

## How did you run restic exactly?

```
export B2_ACCOUNT_ID=...
export B2_ACCOUNT_KEY=...
export RESTIC_REPOSITORY=b2:lars-backup-home
restic snapshots
```

## What backend/server/service did you use to store the repository?

backblaze b2

## Expected behavior

The B2 service seemed to be having some availability issues.  I expected restic to eventually time out.

## Actual behavior

Restic blocked indefinitely.  I cancelled it after 25 minutes.

## Do you have an idea how to solve the issue?

There ought to be conservative default timeouts for most operations, and possibly a `--timeout` option that would permit the operator to modify the default timeout values.

**Comment by stevesbrain** [comment]
I may not have been patient enough, but I've seemed to have a similar issue when I lose Wi-Fi, for example (or I pull the Ethernet cord). I've watched for approximately 10 minutes (again, not a measured time; just roughly what it felt like), and it never timed out. Is it supposed to after a certain amount of time? If so, I'll test it out :)

**Comment by fd0** [comment]
Hm. I'm not so sure what we should do here, but we can discuss in this issue. @kurin said on IRC that B2 requires the client to handle all errors, usually by being patient enough and retrying as long as it takes.

**Comment by kurin** [comment]
As @larsks pointed out in IRC, at the very least it makes sense for interactive commands to have a short timeout, which is easy enough to do.  The B2 lib will return immediately once its context is cancelled, so that's easy enough to do.

Disconnecting and reconnecting to the network should result in an eventual resumption of service; if you can reproduce that and get a debug log I'd be grateful.

**Comment by stevesbrain** [comment]
@kurin I went to get that debug log for you by backing up to rest server, remote, then `iptables` dropping the traffic, waiting 10-15 minutes, then deleting that rule + waiting to see if it resumed. It did actually resume, so I obviously just wasn't waiting long enough previously!

Tried the same with b2 as the destination, and again, it actually resumed fine :) All good on this end - I'll keep the debug logs in case you're interested nonetheless.
